### PR TITLE
refactor: remove application service dependency on secret service

### PR DESCRIPTION
**Pull request by wallyworld**
The secret service was being passed as a dependency to the application service, to allow the application service to delete app and unit owned secrets. Instead of constructing the entire secret service, pass in just the components needed for the application service to construct a secret deleter to do the job.

Since the secret service is not used directly by the application service, we can move the state logic to get secret owners from secret to application domain. This change to state is literally cut and paste.

For secret deletion, a SecretDeleter is created by both the application and secret services - this contains the complicated logic needed to delete secrets and maintain backends etc. The application service is now constructed with just the bits needed for the secret deleter, not the entire secret service. The components needed are a backend reference deleter and a backend admin config getter for a model.

There's still work needed to move some existing params out of the service packages, but this is for another PR.

As a drive by, fix an issue deleting secrets - the ports domain work neglected to update unit deleting to account for deleting the unit endpoint rows. Add this and update the test.

Also a drive by fix to the shell tests because "microk8s" is a valid bootstrap provider.

## QA steps

juju bootstrap microk8s
juju add-model test
juju deploy snappass-test
juju exec -u snappass-test/0 -- secret-add foo=bar
juju remove-application snappass-test

## Links

**Jira card:** [JUJU-6930](https://warthogs.atlassian.net/browse/JUJU-6930)

[JUJU-6930]: https://warthogs.atlassian.net/browse/JUJU-6930?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: domain/application/service/application.go [hunk 1]
```diff
@@ -202,22 +215,17 @@ var (
...
+	DeleteSecret(ctx domain.AtomicContext, uri *coresecrets.URI, params secretservice.DeleteSecretParams) (func(context.Context), error)
 }
 
-// NotImplementedSecretService defines a secret service which does nothing.
-type NotImplementedSecretService struct{}
-
-func (s NotImplementedSecretService) GetSecretsForOwners(ctx domain.AtomicContext, owners ...secretservice.CharmSecretOwner) ([]*coresecrets.URI, error) {
-	return nil, nil
-}
+// NotImplementedSecretDeleter defines a secret service which does nothing.
```
Comment (SimonRichardson) [inline domain/application/service/application.go:224]: This isn't a service.

File: domain/application/service/application.go [hunk 2]
```diff
@@ -227,26 +235,39 @@ type ApplicationService struct {
...
 	}
-	if params.Secrets == nil {
-		params.Secrets = NotImplementedSecretService{}
+	if params.BackendAdminConfigGetter == nil {
+		params.BackendAdminConfigGetter = secretservice.NotImplementedBackendConfigGetter
+	}
+	if params.SecretBackendReferenceDeleter == nil {
+		params.SecretBackendReferenceDeleter = NotImplementedSecretDeleter{}
+	}
+	secretDeleter := &secretservice.SecretDeleter{
```
Comment (SimonRichardson) [inline domain/application/service/application.go:258]: This doesn't seem right. Why are we reliant on a service here? If it truly is a component, then the service can define an interface and operate on that. Then you could offer an implementation from the secret domain that provided that or have an application only version. Either way, we're coupling services here.

File: domain/application/state/application.go [hunk 3]
```diff
@@ -369,6 +371,55 @@ func (st *ApplicationState) deleteSimpleApplicationReferences(ctx context.Contex
...
+// GetSecretsForOwners returns the secrets owned by the specified apps and/or units.`
+func (st *ApplicationState) GetSecretsForOwners(
+	ctx domain.AtomicContext, appOwners domainsecret.ApplicationOwners, unitOwners domainsecret.UnitOwners,
+) ([]*secrets.URI, error) {
+	if len(appOwners) == 0 && len(unitOwners) == 0 {
+		return nil, errors.New("must supply at least one app owner or unit owner")
+	}
+
+	query := `
+SELECT sm.secret_id AS &secretID.id
```
Comment (manadart) [inline domain/application/state/application.go:383]: I'd create a view for this and use it in both places.

**Review by manadart — COMMENTED** [review]
This intricately couples the two services. I don't think we should pursue this pattern.

**Comment by wallyworld** [comment]
> This intricately couples the two services. I don't think we should pursue this pattern.

We weren't happy with the previous pattern either. We need a solution for this problem. Services need to be able to collaborate as part of a workflow. If we don't want to use the dependency engine to wire up dependent services, and we don't want to use composition, then we need something.

File: domain/application/service/application.go (hunk 2 above)
Comment (wallyworld) [reply domain/application/service/application.go:258]: There's complex logic in deleting a secret that is rightly implemented solely in the secret service. Services should be able to collaborate as part of a workflow. Either we allow services to be wired together via the dependency engine, we allow composition to be used, or we accept we need to cut and paste non trivial logic in different domains.

File: domain/application/state/application.go (hunk 3 above)
Comment (manadart) [reply domain/application/state/application.go:383]: I was looking at this today and realised that this is a _movement_ of the service from secret to application. Can you propose a patch with just that part?

**Pull request closed without merging**